overlord/devicestate/devicemgr.go

```diff
@@ -1802,50 +1802,73 @@ func (m *DeviceManager) Systems() ([]*System, error) {
 	return systems, nil
 }
 
-// SystemAndGadgetInfo return the system details including the model
-// assertion and gadget details for the given system label.
-func (m *DeviceManager) SystemAndGadgetInfo(wantedSystemLabel string) (*System, *gadget.Info, error) {
+// SystemAndGadgetAndEncryptionInfo return the system details
+// including the model assertion, gadget details and encryption info
+// for the given system label.
+func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel string) (*System, *gadget.Info, *EncryptionSupportInfo, error) {
 	if m.isClassicBoot {
-		return nil, nil, fmt.Errorf("cannot get model and gadget information on a classic boot system")
+		return nil, nil, nil, fmt.Errorf("cannot get model and gadget information on a classic boot system")
 	}
 	// get current system as input for loadSeedAndSystem()
 	systemMode := m.SystemMode(SysAny)
 	currentSys, _ := currentSystemForMode(m.state, systemMode)
 
 	s, sys, err := loadSeedAndSystem(wantedSystemLabel, currentSys)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// 2. get the gadget volumes for the given seed-label
 	perf := &timings.Timings{}
 	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeGadget}, perf); err != nil {
-		return nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
 	}
 	gadgetSnap := s.EssentialSnaps()[0]
 	if gadgetSnap.Path == "" {
-		return nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
+		return nil, nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
 	}
 	snapf, err := snapfile.Open(gadgetSnap.Path)
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
 	}
 	gadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
 	}
 	gadgetInfo, err := gadget.InfoFromGadgetYaml(gadgetYaml, sys.Model)
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+	}
+
+	// need kernel details to validate storage against encryption
+	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeKernel}, perf); err != nil {
```

Comment:
I missed this, this is allowed but  we probably want to ask for both in a single LoadEssentialMeta above. The order in EssentialSnaps is the same as for asserts.Model.EssentialSnaps

Reply:
That's a nice suggestion, thank you!